python/cudf/cudf/core/series.py

```diff
@@ -5216,3 +5218,47 @@ def isclose(a, b, rtol=1e-05, atol=1e-08, equal_nan=False):
         result_col[equal_nulls] = True
 
     return Series(result_col, index=index)
+
+
+def factorize(values, sort=False, na_sentinel=-1):
+    """Encode the input values as integer labels
+
+    Parameters
+    ----------
+    values: Series, Index, or CuPy array
+        The data to be factorized.
+    na_sentinel : number
+        Value to indicate missing category.
+
+    Returns
+    --------
+    (labels, cats) : (Series, Series)
+        - *labels* contains the encoded values
+        - *cats* contains the categories in order that the N-th
+            item corresponds to the (N-1) code.
```

Comment:
The `cats` return type is input dependent, no? If an Index is input I think it's expected to get an Index in return for example.

Reply:
ohh! I did not notice that at all. Good catch

Reply:
I looked into this and I wanted to make sure the following makes sense. In pandas, factorizing a `pd.Series` or `pd.Index` object gives us back a `pd.Index` here, whereas factorizing an numpy array just returns a numpy array. Would we want to return a *cupy* array for cupy input then? Would we want to return a cupy array for `labels` as well? (currently a `cudf.Series` presumably to avoid a host copy)

Reply:
For `labels` does Pandas return a `pd.Index` object as well?

I think it makes sense to follow Pandas here, where cudf inputs --> cudf index, cupy inputs --> cupy array.

Reply:
For `labels`, pandas returns a numpy array in all three cases - Series, Index, and NumPy array. So we'd return a `cupy` array here I assume? In any event, since `factorize` is a user-facing API, would this warrant the `breaking` label?

Reply:
> For `labels`, pandas returns a numpy array in all three cases - Series, Index, and NumPy array. So we'd return a `cupy` array here I assume?

Yes, this makes sense to me assuming it's always returning integers.

> Agreed re: your second point. Since `factorize` is a user-facing API, would this warrant the `breaking` label?

Yup, this would be a potentially breaking change.